File: src/main/java/no/nav/veilarbaktivitet/aktivitetskort/Aktivitetskort.java
package no.nav.veilarbaktivitet.aktivitetskort;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import lombok.Builder;
import lombok.Data;
import lombok.Singular;
import lombok.With;
import no.nav.veilarbaktivitet.aktivitet.domain.Ident;
import no.nav.veilarbaktivitet.aktivitet.domain.AktivitetStatus;
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Attributt;
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Etikett;
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.LenkeSeksjon;
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Oppgaver;
import no.nav.veilarbaktivitet.aktivitetskort.util.ZonedOrNorwegianDateTimeDeserializer;

import java.time.LocalDate;
import java.time.ZonedDateTime;
import java.util.List;
import java.util.UUID;

@Data
@Builder(toBuilder = true)
@With
public class Aktivitetskort {

    @JsonProperty(required = true)
    UUID id;

    @JsonProperty(required = true)
    String personIdent;

    @JsonProperty(required = true)
    String tittel;

    String beskrivelse;

    @JsonProperty(required = true)
    AktivitetStatus aktivitetStatus;

    LocalDate startDato;
    LocalDate sluttDato;

    @JsonProperty(required = true)
    Ident endretAv;

    @JsonProperty(required = true)
    @JsonDeserialize(using = ZonedOrNorwegianDateTimeDeserializer.class)
    ZonedDateTime endretTidspunkt;

    @JsonProperty(required = true)
    boolean avtaltMedNav;

    Oppgaver oppgave;

    @Singular("handling")
    List<LenkeSeksjon> handlinger;

    @Singular("detalj")
    List<Attributt> detaljer;

    @Singular("etikett")
    List<Etikett> etiketter;
}

File: src/main/kotlin/no/nav/veilarbaktivitet/aktivitetskort/dto/Aktivitetskort.kt
package no.nav.veilarbaktivitet.aktivitetskort.dto

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import jakarta.validation.constraints.Size
import lombok.Singular
import no.nav.veilarbaktivitet.aktivitet.domain.Ident
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Attributt
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Etikett
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.LenkeSeksjon
import no.nav.veilarbaktivitet.aktivitetskort.dto.aktivitetskort.Oppgaver
import no.nav.veilarbaktivitet.aktivitetskort.util.ZonedOrNorwegianDateTimeDeserializer
import java.time.LocalDate
import java.time.ZonedDateTime
import java.util.*

data class Aktivitetskort(
    @JsonProperty(required = true)
    val id: UUID,
    @JsonProperty(required = true)
    val personIdent: String,
    @JsonProperty(required = true)
    @Size(max = 255, message = "Maks lengde på tittel i et aktivitetskort er 255")
    val tittel: String,
    val beskrivelse: String?,
    @JsonProperty(required = true)
    val aktivitetStatus: AktivitetskortStatus,
    val startDato: LocalDate? = null,
    val sluttDato: LocalDate? = null,
    @JsonProperty(required = true)
    val endretAv: Ident,
    @JsonProperty(required = true)
    @JsonDeserialize(using = ZonedOrNorwegianDateTimeDeserializer::class)
    val endretTidspunkt: ZonedDateTime,
    @JsonProperty(required = true)
    val avtaltMedNav: Boolean,
    val oppgave: Oppgaver? = null,
    @Singular("handling")
    val handlinger: List<LenkeSeksjon>? = null,
    @Singular("detalj")
    val detaljer: List<Attributt>? = null,
    @Singular("etikett")
    val etiketter: List<Etikett>? = null
) {}
